### ISSUE TITLE
CSRF errors are not handled well and lead to data loss

### ISSUE THREAD

**Issue by spantaleev** [zadam/trilium#541]
I'm trying to set up synchronization between a server (localhost, for testing) and the Electron desktop app.

The server is using the Docker image: `0.31-latest`

The desktop is installed from AUR: https://aur.archlinux.org/packages/trilium-bin/

According to **About Trilium Notes**, both the server and client are the same version: 0.31.6, build: 81d10fa605cb85f88554c7e5136d24e5fbe2263d

Both server and client start with an empty data directory.

----

# Rough steps to reproduce

1. Start the server: `docker run -it --rm -p 8080:8080 -v ~/Desktop/trilium-data:/root/trilium-data zadam/trilium:0.31-latest`

2. Go to http://localhost:8080 and create a new user

3. Start the Electron app from the console (`trilian`). We do this, so we can look at the output.

4. Set up sync with the server

5. Log in to the server's web interface as well

6. Click around (expanding subtrees, creating new notes, editing existing notes) in both Electron and the server's web interface

Soon enough you'll see Electron spitting out errors like this:

```
{ ForbiddenError: invalid csrf token
    at csrf (/opt/trilium/resources/app.asar/node_modules/csurf/index.js:112:19)
    at Layer.handle [as handle_request] (/opt/trilium/resources/app.asar/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/trilium/resources/app.asar/node_modules/express/lib/router/route.js:137:13)
    at checkApiAuth (/opt/trilium/resources/app.asar/src/services/auth.js:37:9)
    at Layer.handle [as handle_request] (/opt/trilium/resources/app.asar/node_modules/express/lib/router/layer.js:95:5)
    at next (/opt/trilium/resources/app.asar/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/opt/trilium/resources/app.asar/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/opt/trilium/resources/app.asar/node_modules/express/lib/router/layer.js:95:5)
    at /opt/trilium/resources/app.asar/node_modules/express/lib/router/index.js:281:22
    at param (/opt/trilium/resources/app.asar/node_modules/express/lib/router/index.js:354:14) code: 'EBADCSRFTOKEN' }
```

# Observations

- CSRF checks cause trouble

- pressing the **Today** button silently fails - no error message displayed or anything

- when expanding a note subtree, a CSRF check failure just leaves the progress spinner spinning forever. No error is shown or anything. I'd expect 2 things to happen instead:
  * the UI would not get stuck, but rather get restored to what it was before the action
  * an error message would be shown in the UI (I currently need to look at the console to figure out that a CSRF error happened)

![trilium-spinning-forever](https://user-images.githubusercontent.com/388669/58462826-54f66c00-813b-11e9-8cf1-2d7f184622f8.png)

- editing a note is even worse - changes will not be preserved and you won't even know, because the error is not shown in the UI. Switching to another note and coming back to the one you edited will erase your changes locally too. From what I see, it doesn't seem like Trilium preserves note changes locally and will retry syncing later. It's **eating your data silently** :x:

- going to **Options** -> **Advanced** and pressing **Force full sync** or **Fill sync rows** shows some nice green messages, even though it also experiences CSRF errors. Those errors are not displayed anywhere in the UI.

- going to **Options** -> **Sync** and pressing **Test sync** shows a CSRF failure message as expected :heavy_check_mark: 

- there's no Logout button in the Electron app, which I find a bit weird. Looks like it's mentioned in [this comment](https://github.com/zadam/trilium/issues/311#issuecomment-452742990) ("it doesn't make sense on Electron"). If not Logout, then something similar maybe.. Since this is a session synced to a server (master), I'd expect to be able to "destroy the local DB and start over", which may be why I was looking for a Logout button. For completely-local sessions, I do agree that apps don't usually have a Logout feature. Still, that kind of local apps also don't offer a "create a new user + password" or "log in to some server" options too.

- I also observe some duplicates on both the server and client

Looks like synchronization may be getting through after some reloads (`Ctrl + R`).

![trilium-duplicates](https://user-images.githubusercontent.com/388669/58464478-6db45100-813e-11e9-9a40-1efb9b8fa7a7.png)

# Logs

- [trilium-electron-log.txt](https://github.com/zadam/trilium/files/3226398/trilium-electron-log.txt)
- [trilium-docker-log.txt](https://github.com/zadam/trilium/files/3226399/trilium-docker-log.txt)

**Comment by zadam** [comment]
Hello, I'm sorry you're experiencing these issues. They all seem to originate from the CSRF problem which I'm struggling to reproduce/identify to be able to fix it. You have some valid notes about incorrect error handling, but I consider these lower priority problems since this CSRF problem just should not happen at all ...

> If not Logout, then something similar maybe.. Since this is a session synced to a server (master), I'd expect to be able to "destroy the local DB and start over", which may be why I was looking for a Logout button.

I guess this would be possible, but it's pretty drastic (and potentially data-losing action) that it's not supported in the UI and you need to manually delete the local `trilium-data` folder.

> For completely-local sessions, I do agree that apps don't usually have a Logout feature.

There are two things:

* you can create purely local session to the app - here the login/logout doesn't make sense for desktop app since I think the user is "authenticated/authorized enough" by the OS to access the local data
* you can establish sync to the remove sync server - here is one time authentication to the sync server after which complete "document" is synchronized locally. After that happens, both sync server and desktop instance(s) have identical / interchangeable documents and sync is using secrets stored in the document to authenticate the sync so it's not really possible to "log out" - i.e. to remove sync server authentication credentials because it would remove its own credentials as well. But you can easily break the sync by changing the sync url for example. If you point is to just reset your instance, then again you can just delete the `trilium-data` directory.

> Still, that kind of local apps also don't offer a "create a new user + password" or "log in to some server" options too.

Trilium can be used purely locally or on the web or mix of both and we don't do any distinction between that on the DB level, that's why we require username/password also for initial installation on the desktop. Password is also required as the encryption key for the protected notes.

Since Trilium is single user, username is kind of unnecessary, but I chose to include it anyway as it's common UX pattern ...

**Comment by zadam** [comment]
I published 0.32.2-beta which contains some extra CSRF related logging - could you try it and see if the CSRF problem is reproducible? If so, the logs might shed some light on why this is happening ... thanks!

**Comment by spantaleev** [comment]
Sorry, I haven't had the time to play with this yet.
I'll give it a try soon and report back!

**Comment by spantaleev** [comment]
I've done some tests.

When not using Electron, everything seems fine.
I've worked with 2 Docker servers like this:

```
docker network create trilium

# Server-1
docker run -it --rm -p 8081:8080 --network=trilium --name=trilium-1 -v ~/Desktop/trilium-data-1:/root/trilium-data zadam/trilium:0.32.2-beta

# Server 2 (running a patched image)
docker run -it --rm -p 8082:8080 --network=trilium --name=trilium-2 -v ~/Desktop/trilium-data-2:/root/trilium-data zadam/trilium:0.32.2-beta-patched
```

server-2 needs to run a patched image, as described in #551. Otherwise the "sync with another server" option is not available in the UI.

Access to the web UI is done via the public ports:
- `http://localhost:8081` (trilium-1)
- `http://localhost:8082` (trilium-2)

Setting up sync from `trilium-2` to `trilium-1` can happen using a Sync URL that would work across the Docker network (`http://trilium-1:8080`).

Both servers get their own CSRF token with a different secret. Synchronization works okay in both directions, except for some minor other bugs I've found: #551, #552.

--------

The problem described in this issue seems to happen when Electron is used.

> Generated CSRF token BmsHcic5-oYlheqVZOraQor3uxtAFVGRnxaw with secret undefined

The secret being `undefined` happens each and every time. I guess that may be what's causing trouble.

I think I saw secret being `undefined` once, when using the web client (not Electron), but I haven't been able to reproduce it since then. Perhaps the web client was loaded from an old version of Trilium, or had some past login state, etc.

**Comment by zadam** [comment]
Thank you for further testing, this is very helpful. I'm trying to dig deeper, but so far no luck in reproducing the issue.

There should be also "Invalid CSRF token: " errors in the log, could you post here an example? Alternatively attaching the whole log file would be great!

**Comment by spantaleev** [comment]
Here are the full logs:

- [trilium-0.32.2-beta-server.log](https://github.com/zadam/trilium/files/3244587/trilium-0.32.2-beta-server.log)
- [trilium-0.32.2-beta-electron.log](https://github.com/zadam/trilium/files/3244588/trilium-0.32.2-beta-electron.log)

The steps are like this:
- launch the server using a Docker container: `docker run -it --rm -p 8081:8080 --network=trilium --name=trilium-1 -v ~/Desktop/trilium-data-1:/root/trilium-data zadam/trilium:0.32.2-beta`
- go to `http://localhost:8081` and set up the server with a new username + password
- log in to the server using the same username + password
- leave the server's web client open without doing any actions
- now start the Electron client (`trilium`)
- log in to the server (`http://localhost:8081`) with the same username and password
- leave the Electron client open without doing any actions

The Electron dev console also contains this:

```
Created note tab v9Cb / tab_context.js:122
Switched tab v9Cb to root / note_detail.js:385
Activated tab v9Cb / tree_cache.js:39

Uncaught (in promise) TypeError: noteRows is not iterable
    at TreeCache.addResp (tree_cache.js:39)
    at TreeCache.getNotes (tree_cache.js:101)
```

**Comment by zadam** [comment]
I finally managed to reproduce the problem! Short story - the problem lies in the fact that you're running both instances on the same host (localhost). Trilium frontend process then connects to two instances on the same host - one for getting regular data, the other syncing from the server which means they are getting two different CSRF secrets - and the sync server one overwrites the local one so it then can't be used to authenticate against sync servers. The fact that instances run on different ports doesn't help since cookies are shared for all ports on the same host.

This was difficult for me to reproduce because I have set up different hosts for testing of sync so cookies did not conflict. 

This situation should arise in testing only since it otherwise doesn't make much sense to have such setup in "production". But I'll check if it's possible to fix this by removing CSRF check for sync requests (where it's not necessary anyway ...)
